### PR TITLE
[SW-1323] Expose 'modelId' as a parameter of H2O algorithms

### PR DESCRIPTION
**Pull request by mn-mikke**
The 'modelId' parameter will be exposed on:
* H2OGBM
* H2OGLM
* H2OXGBoost
* H2ODeepLearning

### CODE REVIEW

File: ml/src/test/scala/org/apache/spark/ml/spark/models/H2OAlgoTest.scala [hunk 1]
```diff
@@ -58,6 +60,45 @@ class H2OAlgoTest extends FunSuite with SharedH2OTestContext {
     loadedModel.transform(dataset).count()
   }
 
+  test("H2OGLM with set modelId is trained mutliple times")
+  {
```
Comment (jakubhava) [inline ml/src/test/scala/org/apache/spark/ml/spark/models/H2OAlgoTest.scala]: NIT: any reason why `{` is on the new line ?

File: ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OGBM.scala [hunk 2]
```diff
@@ -29,7 +29,11 @@ class H2OGBM(override val uid: String) extends H2OAlgorithm[GBMParameters] with
 
   def this() = this(Identifiable.randomUID("gbm"))
 
-  override def trainModel(params: GBMParameters): GBMModel = new GBM(params).trainModel().get()
+  override def trainModel(params: GBMParameters): GBMModel = {
+    val modelId = getModelId()
+    val builder = if (modelId == null || modelId.isEmpty) new GBM(params) else new GBM(params, createKey(modelId))
```
Comment (jakubhava) [inline ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OGBM.scala]: This code is duplicated across all H2OAlgorithms. I can see that we create different algo param instance in each, but what do you think about encapsulating this functionality into a method  which takes type parameter( such as `GBM`) and can create new instance using reflection?

File: ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OAlgorithm.scala [hunk 3]
```diff
@@ -85,4 +85,20 @@ abstract class H2OAlgorithm[P <: Model.Parameters : ClassTag] extends Estimator[
   }
 
   override def copy(extra: ParamMap): this.type = defaultCopy(extra)
+
+  protected def createKey[T <: Keyed[T]](modelId: String): Key[T] = {
+    val key = Key.make[T](modelId)
+    if(H2O.containsKey(key)) {
```
Comment (jakubhava) [inline ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OAlgorithm.scala]: NIT: formatting - missing space after`if`

File: ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OAlgorithm.scala [hunk 4]
```diff
@@ -85,4 +85,20 @@ abstract class H2OAlgorithm[P <: Model.Parameters : ClassTag] extends Estimator[
...
+      var i = 0
```
Comment (jakubhava) [inline ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OAlgorithm.scala]: what do you think of renaming `i` to something like `suffix`?

File: ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OAlgorithm.scala [hunk 5]
```diff
@@ -85,4 +85,20 @@ abstract class H2OAlgorithm[P <: Model.Parameters : ClassTag] extends Estimator[
...
+      var replacement = key
+      do {
```
Comment (jakubhava) [inline ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OAlgorithm.scala]: Just nit: Maybe we could expose the replacement functionality to a new, small private method and keep this method as:

if(keyExists){
 createReplacement()
} else {
key}

File: ml/src/test/scala/org/apache/spark/ml/spark/models/H2OAlgoTest.scala (hunk 1 above)
Comment (mn-mikke) [reply ml/src/test/scala/org/apache/spark/ml/spark/models/H2OAlgoTest.scala]: There is no reason :-) I will fix it, thanks!

File: ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OGBM.scala (hunk 2 above)
Comment (mn-mikke) [reply ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OGBM.scala]: Yeah, that' a great idea! Will do, thanks!

File: ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OAlgorithm.scala (hunk 4 above)
Comment (mn-mikke) [reply ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OAlgorithm.scala]: Yup, no problem :-)

**Comment by mn-mikke** [comment]
@jakubhava Thank you for your valid comments! Please can you review the PR one more time?

**Pull request merged**